<a href="https://colab.research.google.com/github/valperezg02/Analitica-de-Negocios/blob/main/Reto_2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##**Caso de Estudio**

Un centro de salud quiere implementar un modelo de árbol de decisión para predecir la probabilidad de qué personas van a desarrollar una enfermedad cardiaca

Para este proceso se usarán las siguientes variables:

* Age: Indica el número de años que posee la persona. Entre más alto el numero, más probabilidad tiene de entrar en el análisis.
* Systolic: Indica la presión arterial máxima, altos valores en esta variable indican problemas relacionados con enfermedad cardiaca.
* Diastolic: Indica la presión arterial mínima, altos valores en esta variable también indican problemas relacionados con enfermedad cardiaca.
* BMI: Indica el índice de masa corporal, entre más alta sea esta variable, indica obesidad o sobrepeso, lo que significa alta probalidad de sufrir de una enfermedad cardiaca.

Se quiere clasificar las personas según la probabilidad de sufrir de una enfermedad cardiaca


0. Se inicia con la carga de las librerias

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

1. Se carga la base de datos HeartDisease

In [ ]:
nxl = '/content/2. HeartDisease.xlsx'
XDB = pd.read_excel(nxl,sheet_name=0)

XD = XDB.iloc[:,[1,2,3,4]]
yd = XDB.iloc[:,0]
display(XD)
display(yd)


,Age,Systolic,Diastolic,BMI
0,44,112,111,17
1,55,128,90,27
2,47,131,94,26
3,31,151,104,17
4,65,148,117,17
...,...,...,...,...
175,43,119,76,25
176,63,107,113,40
177,47,116,61,27
178,70,110,88,24


,Disease
0,0
1,1
2,1
3,0
4,1
...,...
175,0
176,1
177,0
178,0


2. Se implementa el modelo de Árbol de decisión

In [ ]:
mar = DecisionTreeClassifier(criterion='gini', max_depth=4)
mar.fit(XD,yd)

ydp = mar.predict(XD)

cm = confusion_matrix(yd,ydp)
display(cm)
VN = cm[0,0]; FP = cm[0,1]; FN = cm[1,0]; VP = cm[1,1]

array([[63, 10],
       [20, 87]])

**Análisis de Resultados**

De la base de datos se puede ver un total de 73 personas que no desarrollan enfermedad cardíaca, mientras que para la categoría de personas que sí desarrollan enfermedad cardíaca la matriz cuenta con un total de 107 personas.

De los 73 datos inciales, este modelo pronosticó correctamente un total de 63 personas.

De los 107 datos que indican personas con enfermedad cardíaca, el modelo pronosticó correctamente un total de 87 personas.

De acuerdo a lo anterior:
* Este modelo pudo identificar en un 86,3% las personas que no desarrollan una enfermedad cardiaca.
* Mientras que el modelo identificó en un 81,3% a las personas que sí desarrollan enfermedad cardíaca

In [ ]:
# 1. Exactitud
Ex = (VP+VN)/len(XD)
print('La Exactitud es: ', Ex)

# 2. Sensibilidad
SEN = VP/(VP+FN)
print('La Sensibilidad es: ', SEN)

# 3. Especificidad
ESP = (VN/(VN+FP))
print('La Especificidad es: ', ESP)

# 4. Precisión
Pre = VP/(VP+FP)
print('La Precisión es: ', Pre)

# 5. Predicción Negativa
PreNeg = VN/(VN+FN)
print('La Predicción Negativa es: ', PreNeg)

La Exactitud es:  0.8333333333333334
La Sensibilidad es:  0.8130841121495327
La Especificidad es:  0.863013698630137
La Precisión es:  0.8969072164948454
La Predicción Negativa es:  0.7590361445783133


**Análisis de Resultados**

De acuerdo a los resultados de las métricas, es posible decir que el modelo alcanzó una EXACTITUD de 83,33%, esto muestra un buen comportamiento en el proceso de clasificación de pacientes que desarrollan o no una enfermedad cardiaca.

Por otro lado, la ESPECIFICIDAD fue de 86,30%, esto muestra la gran capacidad que tiene el modelo para identificar de manera correcta a las persona que no desarrollan una enfermedad cardiaca, de esta manera se reduce los diagnósticos innecesarios.

La PRECISIÓN tuvo un 89,69%, lo que indica que la predicción del modelo acerca de las personas que desarrollan una enfermedad cardiaca es correcta.

La SENSIBILIDAD fue de 81,31%, esto muestra que el modelo puede detectar efectivamente las personas que desarrollan la enfermedad.

La PREDICCIÓN NEGATIVA es de 75,90%, lo que muestra que el modelo si indica de manera correcta a las personas como sanas.

Como conclusión se puede decir que el modelo está por encima de 75% lo que indica un buen despeño y buenos resultados para identificar las personas que desarrollan una enfermedad cardiaca.


3. Se despliega el árbol de decisión

In [ ]:
from sklearn.tree import export_graphviz
from pydotplus import graph_from_dot_data

vs = ['Age', 'Systolic', 'Diastolic', 'BMI']
dot_data = export_graphviz(mar,feature_names=vs)
graph = graph_from_dot_data(dot_data)
graph.write_png('Arbol.png')

True

**Análisis de Resultados**

De acuerdo al árbol que otrogó Colab, la varibale más importante fue AGE, porque es el nodo raíz en el modelo. Esto quiere decit que la probabilida de desarrollar la enfermedad cardiaca esta ligada por el el factor de la edad.

Luego se incorpora l avariable BMI, Systolic y Diastolic, lo que dice que el modelo tambien considera importante estas variables para determinar si se desarrollar una enfermedad cardiaca.

Se pueden evidenciar varios nodos puros, lo que quiere decir que el modelo fue capaz de clasificar correctamente algunso grupos.

Esto evidencia que las edad y la presión elevada incrementa la probabilidad de desarrollar una enfermedad cardiaca.

En conclusión, este modelo tiene reglas claras, que se pueden interpretar fácilmente para identificar el riesgo.